src/Simulation/Simulators/OpenSystemsSimulator/Decompositions/IntrinsicInterface.qs

```diff
@@ -0,0 +1,210 @@
+        body (...) {
+            Native.S(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                S(target);
+            } elif (Length(controls) == 1) {
+                T(controls[0]);
+                T(target);
+                within {
+                    Controlled X([controls[0]], target);
+                } apply {
+                    Adjoint T(target);
+                }
+            } else {
+                ApplyWithLessControlsA(Controlled S, (controls, target));
+            }
+        }
+    }
+
+    operation T(target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            Native.T(target);
+        }
+
+        controlled (controls, ...) {
+            if Length(controls) == 0 {
+                T(target);
+            } else {
+                // TODO: Decompositions of `Controlled T` currently used in Q#
+                //       target packages rely on R1Frac, which is not yet
+                //       implemented in experimental simulation.
+                fail "Controlled T operations are not yet supported.";
+            }
+        }
+    }
+
+    // NB: We separate out this operation to avoid hardware targeting rewrite
+    //     steps from trying to lift this operation and modifying the C#
+    //     code gen in the process.
+    internal operation MeasureWithoutPauliI(bases : Pauli[], register : Qubit[]) : Result {
+        mutable newBases = [];
+        mutable newQubits = [];
+        // NB: using Zipped would be nice here, but this is built before
+        //     M.Q.Standard...
+        for idxBasis in 0..Length(bases) - 1 {
+            if bases[idxBasis] != PauliI {
+                set newBases += [bases[idxBasis]];
+                set newQubits += [register[idxBasis]];
+            }
+        }
+
+        if Length(newBases) == 0 {
+            return Zero;
+        } else {
+            return Measure(newBases, register);
+        }
+    }
+
+    operation Measure(bases : Pauli[], register : Qubit[]) : Result {
+        // If anything is PauliI, strip out and recurse.
+        if IsAnyPauliI(bases) {
+            return MeasureWithoutPauliI(bases, register);
+        }
+
+        // At this point, we're guaranteed that bases contains no PauliI, and
+        // has length at least one. The simulator implementation likewise
+        // guarantees that in the single-qubit case, the basis is not Z.
+        // Let's check the easy single-qubit case first, then, and handle the
+        // X and Y cases.
+        if Length(bases) == 1 {
+            let basis = bases[0];
+            let target = register[0];
+            mutable result = Zero;
+            within {
+                if basis == PauliX {
+                    Native.H(target);
+                } elif basis == PauliY {
+                    Native.H(target);
+                    Native.S(target);
+                } else {
+                }
```

Comment:
nitpick: Putting a comment about Pauli-Z in the empty else-block will make the intent a little clearer and the pattern easier to see.

Reply:
Good point, yeah. I'll go on and add that.